# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [1]:
# import numpy and pandas
import pandas as pd
import numpy as np
from scipy.stats import trim_mean, mode, skew, gaussian_kde, pearsonr, spearmanr, beta
from statsmodels.stats.weightstats import ztest as ztest

from scipy.stats import ttest_ind, norm, t
from scipy.stats import f_oneway
from scipy.stats import sem

# Challenge 1 - Exploring the Data

In this challenge, we will examine all salaries of employees of the City of Chicago. We will start by loading the dataset and examining its contents

In [3]:
# Run this code:
salaries = pd.read_csv('C:/Users/anmnu/DSML/Week4/Day18/Lab/lab-hypothesis-testing-en/data/Current_Employee_Names__Salaries__and_Position_Titles.csv')

Examine the `salaries` dataset using the `head` function below.

In [4]:
# Your code here
salaries.head()

,Name,Job Titles,Department,Full or Part-Time,Salary or Hourly,Typical Hours,Annual Salary,Hourly Rate
0,"AARON, JEFFERY M",SERGEANT,POLICE,F,Salary,NaN,101442.0,NaN
1,"AARON, KARINA",POLICE OFFICER (ASSIGNED AS DETECTIVE),POLICE,F,Salary,NaN,94122.0,NaN
2,"AARON, KIMBERLEI R",CHIEF CONTRACT EXPEDITER,GENERAL SERVICES,F,Salary,NaN,101592.0,NaN
3,"ABAD JR, VICENTE M",CIVIL ENGINEER IV,WATER MGMNT,F,Salary,NaN,110064.0,NaN
4,"ABASCAL, REECE E",TRAFFIC CONTROL AIDE-HOURLY,OEMC,P,Hourly,20.0,NaN,19.86


# Challenge 2 - Hypothesis Tests

In this section of the lab, we will test whether the hourly wage of all hourly workers is significantly different from $30/hr. Import the correct one sample test function from scipy and perform the hypothesis test for a 95% two sided confidence interval.

In [5]:
# Your code here
import scipy.stats as st

In [6]:
salaries["Salary or Hourly"].value_counts()

Salary or Hourly
Salary    25161
Hourly     8022
Name: count, dtype: int64

In [12]:
# making a new df just for hourly workers
salaries_hourly = salaries[salaries["Salary or Hourly"] == "Hourly"]
salaries_hourly.info()
print(salaries_hourly["Salary or Hourly"].value_counts())
print(salaries_hourly["Hourly Rate"].value_counts())

<class 'pandas.core.frame.DataFrame'>
Index: 8022 entries, 4 to 33175
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               8022 non-null   object 
 1   Job Titles         8022 non-null   object 
 2   Department         8022 non-null   object 
 3   Full or Part-Time  8022 non-null   object 
 4   Salary or Hourly   8022 non-null   object 
 5   Typical Hours      8022 non-null   float64
 6   Annual Salary      0 non-null      float64
 7   Hourly Rate        8022 non-null   float64
dtypes: float64(3), object(5)
memory usage: 564.0+ KB
Salary or Hourly
Hourly    8022
Name: count, dtype: int64
Hourly Rate
35.60    1415
40.20     781
36.21     626
19.86     418
46.10     306
         ... 
42.25       1
32.59       1
49.52       1
28.00       1
35.43       1
Name: count, Length: 167, dtype: int64


In [17]:
hourly_ttest = st.ttest_1samp(salaries_hourly["Hourly Rate"], 30)
print(hourly_ttest)

TtestResult(statistic=np.float64(20.6198057854942), pvalue=np.float64(4.323024048623005e-92), df=np.int64(8021))


In [16]:
print(hourly_ttest.pvalue)

4.323024048623005e-92


In [19]:
# 95% confidence interval, meaning that alpha = 0.05
# We want to see if our pvalue is less than alpha, if yes, we can reject the null hypothesis
float(hourly_ttest.pvalue) < 0.05

True

* Our pvalue (4.32 E-92) is smaller than 0.05. Therefore, we reject the null hypothesis and say that the hourly rage for the hourly workers is significantly different from 30 per hour. 

In [ ]:
# this one is just for my sanity check
salaries_hourly["Hourly Rate"].describe()

count    8022.000000
mean       32.788558
std        12.112573
min         2.650000
25%        21.200000
50%        35.600000
75%        40.200000
max       109.000000
Name: Hourly Rate, dtype: float64

* My observation is that, here we see the mean being 2.78 dollars per hour greater than 30. We have a verey small p value, a very large data set. The large size leads to high statistical power, making even small deviations from 30 statistically significant. 

# Challenge 3 - Constructing Confidence Intervals

While testing our hypothesis is a great way to gather empirical evidence for accepting or rejecting the hypothesis, another way to gather evidence is by creating a confidence interval. A confidence interval gives us information about the true mean of the population. So for a 95% confidence interval, we are 95% sure that the mean of the population is within the confidence interval. 
).

To read more about confidence intervals, click [here](https://en.wikipedia.org/wiki/Confidence_interval).


In the cell below, we will construct a 95% confidence interval for the mean hourly wage of all hourly workers. 

The confidence interval is computed in SciPy using the `t.interval` function. You can read more about this function [here](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.t.html).

To compute the confidence interval of the hourly wage, use the 0.95 for the confidence level, number of rows - 1 for degrees of freedom, the mean of the sample for the location parameter and the standard error for the scale. The standard error can be computed using [this](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.sem.html) function in SciPy.

In [26]:
# Your code here
# we use st.t.interval(confidence, df, loc=mean, scale=std)
n = len(salaries_hourly) # size of our salary worker dataset
df = n-1 
confidence = 0.95
mean = st.tmean(salaries_hourly["Hourly Rate"]) 
std = st.tstd(salaries_hourly["Hourly Rate"]) # But here we are asked to use standard error sem instead, so
sem = st.sem(salaries_hourly["Hourly Rate"])

print("The mean of the Hourly Rate is: ", mean)
print("The standard error for mean of the Hourly Rate is: ", sem)
confidence_intervals_hourly = st.t.interval(confidence, df, loc = mean, scale = sem)
print("The confidence intervals with 95 percent confidence level: ", confidence_intervals_hourly)


The mean of the Hourly Rate is:  32.78855771628024
The standard error for mean of the Hourly Rate is:  0.1352368565101596
The confidence intervals with 95 percent confidence level:  (np.float64(32.52345834488425), np.float64(33.05365708767623))


* The confidence intervals with 95% confidence level is from 32.523 to 33.054 dollars/hour.

# Challenge 4 - Hypothesis Tests of Proportions

Another type of one sample test is a hypothesis test of proportions. In this test, we examine whether the proportion of a group in our sample is significantly different than a fraction. 

You can read more about one sample proportion tests [here](http://sphweb.bumc.bu.edu/otlt/MPH-Modules/BS/SAS/SAS6-CategoricalData/SAS6-CategoricalData2.html).

In the cell below, use the `proportions_ztest` function from `statsmodels` to perform a hypothesis test that will determine whether the number of hourly workers in the City of Chicago is significantly different from 25% at the 95% confidence level.

In [27]:
from statsmodels.stats.proportion import proportions_ztest

In [29]:
# Your code here
count = len(salaries_hourly) # hourly workers
nobs = len(salaries) # total workers
value = 0.25 # 25% asked by the question

stat, pval = proportions_ztest(count, nobs, value = 0.25)
print(stat, pval)

-3.5099964213703005 0.0004481127249057967


* The stat tells us how far our observed proportion is from 25%, measured in standard errors. Here we have a stat value of -3.51, telling us that our proportion is less than 25%. With more than 3 standard errors, indicating that the observed proportion is less than 25%

* The pval gives us the pvalue. It is 0.000448, smaller than 0.05 (95% confidence level). We reject the null hypothesis. The proportion of hourly workesr is significantly different from 25%